subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/DefaultVersionedComponentChooser.java

```diff
@@ -68,51 +68,55 @@ private boolean isGeneratedModuleDescriptor(ComponentResolveMetadata componentRe
         return componentResolveMetadata.isGenerated();
     }
 
-    public void selectNewestMatchingComponent(Collection<? extends ModuleComponentResolveState> versions, BuildableComponentSelectionResult result, ModuleVersionSelector requested) {
+    public void selectNewestMatchingComponent(Collection<? extends ModuleComponentResolveState> versions, ComponentSelectionContext result, ModuleVersionSelector requested) {
         VersionSelector requestedVersion = versionSelectorScheme.parseSelector(requested.getVersion());
         Collection<SpecRuleAction<? super ComponentSelection>> rules = componentSelectionRules.getRules();
 
+        // Loop over all listed versions, sorted by LATEST first
         for (ModuleComponentResolveState candidate : sortLatestFirst(versions)) {
             MetadataProvider metadataProvider = createMetadataProvider(candidate);
 
             boolean versionMatches = versionMatches(requestedVersion, candidate, metadataProvider);
             if (!metadataProvider.isUsable()) {
+                // maybe we have already determined that this is NOT usable
+                // so we don't need to proceed further
                 applyTo(metadataProvider, result);
                 return;
             }
+
             String version = candidate.getVersion().getSource();
             if (!versionMatches) {
+                // this version doesn't match the selector, we need to go on with the next listed one
                 result.notMatched(version);
                 continue;
             }
 
             ModuleComponentIdentifier candidateIdentifier = candidate.getId();
             boolean accepted = !isRejectedByRules(candidateIdentifier, rules, metadataProvider);
-            if (!metadataProvider.isUsable()) {
-                applyTo(metadataProvider, result);
-                return;
-            }
-
             if (accepted) {
+                // this version is selected AND not rejected by any metadata rule. This is the one
```

Comment:
Again, the code already says this, it just needs to be refactored a bit here and there. These kinds of comments make refactoring harder, because they become out of date quickly.